File: android/app/src/main/java/com/miniclaw/app/EngineClient.kt
package com.miniclaw.app

import android.os.Handler
import android.os.Looper
import com.miniclaw.core.NativeEngine

/**
 * Owns the native engine; marshals worker-thread events onto the main looper.
 *
 * Threading contract (mirrors ios/Miniclaw/Engine/EngineClient.swift):
 *  - all methods are called from the main thread;
 *  - blocking create/destroy run on a background thread, results posted back;
 *  - [UiListener] callbacks always fire on the main thread.
 */
class EngineClient(val workspaceDir: String) {

    interface UiListener {
        fun onToken(text: String)
        fun onToolStart(content: String)
        fun onToolEnd(content: String)
        fun onStatus(text: String)
        fun onError(text: String)
        fun onDone()
    }

    @Volatile var engine: NativeEngine? = null
        private set

    /** True while a create is in flight (or the engine is running). */
    @Volatile var isStarting: Boolean = false
        private set

    @Volatile var startError: String? = null
        private set

    /** Wall-clock ms when the engine last became ready (null if never). */
    @Volatile var readyAt: Long? = null
        private set

    private val main = Handler(Looper.getMainLooper())
    private var listener: UiListener? = null

    /** Fired on the main thread after a start attempt finishes (success or failure). */
    var onEngineState: (() -> Unit)? = null

    fun setUiListener(l: UiListener) {
        listener = l
    }

    /** Start the engine on a background thread (bootstraps the workspace). No-op if already running. */
    fun start() {
        if (isReady() || isStarting) return
        isStarting = true
        startError = null
        Thread {
            try {
                val e = NativeEngine.create(workspaceDir, null)
                engine = e
                readyAt = System.currentTimeMillis()
                main.post {
                    isStarting = false
                    onEngineState?.invoke()
                }
            } catch (t: Throwable) {
                startError = t.message ?: "engine start failed"
                isStarting = false
                main.post {
                    listener?.onError(startError!!)
                    onEngineState?.invoke()
                }
            }
        }.start()
    }

    fun isReady(): Boolean = engine?.isRunning() == true

    fun version(): String = NativeEngine.version()

    /** Current config.yaml contents, or null if none was loaded. Main thread. */
    fun getConfig(): String? = try {
        engine?.getConfig()
    } catch (t: Throwable) {
        null
    }

    /** Write new config YAML and apply it. @return true on success. Main thread. */
    fun setConfig(yaml: String): Boolean = try {
        engine?.setConfig(yaml) ?: false
    } catch (t: Throwable) {
        false
    }

    fun send(sessionId: String, message: String) {
        val e = engine ?: return
        e.sendMessage(sessionId, message, object : NativeEngine.EventListener {
            override fun onEvent(type: String, content: String) {
                when (type) {
                    "token" -> main.post { listener?.onToken(content) }
                    "tool_start" -> main.post { listener?.onToolStart(content) }
                    "tool_end" -> main.post { listener?.onToolEnd(content) }
                    "status" -> main.post { listener?.onStatus(content) }
                    "error" -> main.post { listener?.onError(content) }
                    "done" -> main.post { listener?.onDone() }
                }
            }
        })
    }

    fun setString(section: String, key: String, value: String): Boolean =
        try {
            engine?.setString(section, key, value) ?: false
        } catch (t: Throwable) {
            false
        }

    fun stop() {
        isStarting = false
        readyAt = null
        engine?.destroy()
        engine = null
        main.post { onEngineState?.invoke() }
    }
}

File: ios/Miniclaw/Engine/EngineClient.swift
import Foundation
import MiniclawCore

/// Owns the native engine (libminiclaw_core.dylib).
///
/// Mirrors Android's EngineClient.kt 1:1 in semantics:
///   - `start()` creates the engine on a background thread (workspace
///     bootstrap is not instant); the handle is published on main.
///   - Engine events arrive on a worker thread (the engine's turn fiber);
///     they are marshalled onto the main thread before `onEvent` fires.
///   - `stop()` destroys the engine; the app does this when backgrounded
///     (like Android's onStop) and restarts on foreground.
///
/// Threading contract: all EngineClient methods are called from the main
/// thread. `engine` is only ever written on main; native calls that block
/// (create/destroy) run on a private queue, with results marshalled back.
///
/// Unlike Android there is no JNI bridge library: Swift imports the C ABI
/// directly through the MiniclawCore module (FFI/module.modulemap).
final class EngineClient {

    struct Event {
        /// "status" | "token" | "tool_start" | "tool_end" | "error" | "done"
        let type: String
        let content: String
    }

    /// Invoked on the main thread.
    var onEvent: ((Event) -> Void)?

    private(set) var isReady = false
    private(set) var isStarting = false
    private(set) var startError: String?

    /// mc_engine* — opaque to Swift. Main-thread only.
    private var engine: OpaquePointer?
    /// Bumped by stop(); a create in flight when the app backgrounds is
    /// discarded (its result never belongs to a live session).
    private var session = 0
    private let queue = DispatchQueue(label: "com.miniclaw.engine")
    private let workspaceDir: String

    init(workspaceDir: String) {
        self.workspaceDir = workspaceDir
    }

    // ── lifecycle ───────────────────────────────────────────────────────────

    /// Create and start the engine in the background. Main thread.
    func start() {
        guard engine == nil, !isStarting else { return }
        isStarting = true
        isReady = false
        let dir = workspaceDir
        let s = session
        queue.async { [weak self] in
            let h = mc_engine_create(dir, nil)
            DispatchQueue.main.async {
                guard let self, self.session == s else {
                    if h != nil { mc_engine_destroy(h) }  // stopped mid-create
                    return
                }
                self.isStarting = false
                if h == nil {
                    let err = String(cString: mc_last_error())
                    self.startError = "engine start failed: \(err)"
                    self.onEvent?(Event(type: "error", content: self.startError ?? ""))
                    return
                }
                self.engine = h
                self.isReady = true
                self.startError = nil
                self.onEvent?(Event(type: "status", content: "ready"))
            }
        }
    }

    /// Shut down the engine and release native resources. Idempotent.
    /// Main thread; the (blocking) destroy runs on the private queue.
    func stop() {
        session += 1          // invalidate any create in flight
        isStarting = false
        guard let h = engine else { return }
        engine = nil          // no new sends can see it
        isReady = false
        queue.async {
            mc_engine_destroy(h)
        }
    }

    var isRunning: Bool {
        guard let engine else { return false }
        return mc_is_running(engine) == 1
    }

    static func version() -> String {
        String(cString: mc_version())
    }

    // ── chat ────────────────────────────────────────────────────────────────

    /// Send a user message. Events stream via `onEvent` (main thread), ending
    /// with "done" (or "error"). Turns are serialized by the engine.
    func send(sessionId: String, message: String) {
        guard let engine else { return }
        mc_send_message(engine, sessionId, message, Self.cBridge,
                        Unmanaged.passUnretained(self).toOpaque())
    }

    /// C ABI event bridge. Runs on the engine's worker thread — must be fast
    /// and non-blocking; we only copy strings and hop to main. `user_data`
    /// is an unretained EngineClient (see send()); the app keeps it alive for
    /// the process lifetime, same as Android's listener object.
    private static let cBridge: mc_event_cb = { userData, typePtr, contentPtr in
        guard let userData else { return }
        let client = Unmanaged<EngineClient>.fromOpaque(userData).takeUnretainedValue()
        let type = typePtr.map { String(cString: $0) } ?? ""
        let content = contentPtr.map { String(cString: $0) } ?? ""
        DispatchQueue.main.async {
            client.onEvent?(Event(type: type, content: content))
        }
    }

    // ── config ──────────────────────────────────────────────────────────────

    /// Current config file contents (YAML), or nil if none was loaded.
    func getConfig() -> String? {
        guard let engine else { return nil }
        let ptr = mc_get_config(engine)
        defer { if let ptr { mc_free_string(ptr) } }
        return ptr.map { String(cString: $0) }
    }

    /// Write new config YAML and apply it. true on success.
    func setConfig(_ yaml: String) -> Bool {
        guard let engine else { return false }
        return mc_set_config(engine, yaml) == 0
    }

    /// Set a single setting (section/key/value). true on success.
    func setString(section: String, key: String, value: String) -> Bool {
        guard let engine else { return false }
        return mc_set_string(engine, section, key, value) == 0
    }
}
